libcloud/common/google.py

```
@@ -0,0 +1,467 @@
+            self.token_info = self.auth_conn.get_new_token()
+            self._write_token_info_to_file()
+
+        self.token_expire_time = datetime.datetime.strptime(
+            self.token_info['expire_time'], TIMESTAMP_FORMAT)
+
+        super(GoogleBaseConnection, self).__init__(user_id, key, **kwargs)
+
+    def add_default_headers(self, headers):
+        """
+        @inherits: L{Connection.add_default_headers}
+        """
+        headers['Content-Type'] = "application/json"
+        headers['Host'] = self.host
+        return headers
+
+    def pre_connect_hook(self, params, headers):
+        """
+        Check to make sure that token hasn't expired.  If it has, get an
+        updated token.  Also, add the token to the headers.
+
+        @inherits: L{Connection.pre_connect_hook}
+        """
+        now = datetime.datetime.utcnow()
+        if self.token_expire_time < now:
+            self.token_info = self.auth_conn.refresh_token(self.token_info)
+            self.token_expire_time = datetime.datetime.strptime(
+                self.token_info['expire_time'], TIMESTAMP_FORMAT)
+            self._write_token_info_to_file()
+        headers['Authorization'] = 'Bearer %s' % (
+            self.token_info['access_token'])
+
+        return params, headers
+
+    def encode_data(self, data):
+        """Encode data to JSON"""
+        return json.dumps(data)
+
+    def request(self, *args, **kwargs):
+        """
+        @inherits: L{Connection.request}
+        """
+        # Adds some retry logic for the occasional
+        # "Connection Reset by peer" error.
+        retries = 4
+        tries = 0
+        while tries < (retries - 1):
+            try:
+                return super(GoogleBaseConnection, self).request(
+                    *args, **kwargs)
+            except socket.error as e:
+                if e.errno == 104:
+                    tries = tries + 1
+                else:
+                    raise
```

Please raise the original exception - `raise e`